usr/lib/pkcs11/ep11_stdll/ep11_func.h

```diff
@@ -331,6 +331,99 @@ typedef long (*xcpa_queryblock_t) (unsigned char *blk, size_t blen,
+                                unsigned int query, unsigned int subquery,
+                                uint64_t target);
+
+#ifndef CK_IBM_XCP_HOSTQ_IDX
+
+typedef enum {
+    CK_IBM_XCPQ_API         =  0,  /* API and build identifier     */
+    CK_IBM_XCPQ_MODULE      =  1,  /* module-level information     */
+    CK_IBM_XCPQ_DOMAINS     =  2,  /* list active domains & WK IDs */
+    CK_IBM_XCPQ_DOMAIN      =  3,
+    CK_IBM_XCPQ_SELFTEST    =  4,  /* integrity & algorithm tests  */
+    CK_IBM_XCPQ_EXT_CAPS    =  5,  /* extended capabilities, count */
+    CK_IBM_XCPQ_EXT_CAPLIST =  6,  /* extended capabilities, list  */
+    CK_IBM_XCPQ_AUDITLOG    =  8,  /* audit record or records      */
+    CK_IBM_XCPQ_DESCRTEXT   =  9,  /* human-readable text/tokens   */
+    CK_IBM_XCPQ_EC_CURVES   = 10,  /* supported elliptic curves    */
+    CK_IBM_XCPQ_COMPAT      = 11,  /* domains' compatibility modes */
+    CK_IBM_XCPQ_MAX         = CK_IBM_XCPQ_COMPAT
+} CK_IBM_XCPQUERY_t;
+
+#define CK_IBM_XCP_HOSTQ_IDX  0xff000000  /* host-only queries index, min. */
+typedef enum {
+    CK_IBM_XCPHQ_COUNT    = 0xff000000, /* number of host-query indexes   */
+                                        /* including this type itself     */
+    CK_IBM_XCPHQ_VERSION  = 0xff000001, /* host-specific package version  */
+                                        /* such as packaging library ID   */
+    CK_IBM_XCPHQ_VERSION_HASH = 0xff000002,
+                                        /* assumed-unique identifier of   */
+                                        /* host code, such as version-    */
+                                        /* identifying cryptographic hash */
+                                        /* (library signature field...)   */
+    CK_IBM_XCPHQ_DIAGS    = 0xff000003, /* host code diagnostic level     */
+                                        /* 0 if non-diagnostics host code */
+    CK_IBM_XCPHQ_HVERSION = 0xff000004, /* human-readable host version    */
+                                        /* identification (recommended:   */
+                                        /* UTF-8 string)                  */
+    CK_IBM_XCPHQ_TGT_MODE = 0xff000005, /* host targeting modes           */
+                                        /* returns supported target modes */
+                                        /* as bitmask                     */
+                                        /* if not available only compat   */
+                                        /* target mode is in use          */
+                                        /* See CK_IBM_XCPHQ_TGT_MODES_t   */
+                                        /**/
+    CK__IBM_XCPHQ_MAX = CK_IBM_XCPHQ_TGT_MODE
+} CK_IBM_XCPHQUERY_t;
+
+typedef struct CK_IBM_XCPAPI_INFO {
+    CK_ULONG firmwareApi;
+    CK_ULONG firmwareConfig;          /* truncated firmware hash */
+} CK_IBM_XCPAPI_INFO;
+
+typedef CK_IBM_XCPAPI_INFO    CK_PTR   CK_IBM_XCPAPI_INFO_PTR;
+
+typedef struct CK_IBM_XCP_INFO {
```

Comment:
Yea, good idea ! See my second patch.

Reply:
There are more fields that could be filled with real token data instead of the generic ock "dummy"data: flags, utcTime,  serialNumber, ? compare to CK_TOKEN_INFO definition

Reply:
Regarding Flags, I don't know if this adds anything useful. The wire spec does not say much about the flags they supply.

utcTime is filled when C_GetTokenInfo is called with the current Linux time. The version query is only executed during token initialization, so supplying that UTC time would be wrong.

The serialnumber might be useful, however when multiple adapters are available, I assume that each adapter has its own serial number. So we could simply supply the serial number of the first adapter. Nevertheless, due to hot-plug, it might be that it will display a different serial number dependent on which adapters are online at the time of init.

What do you think?

Reply:
The CK_TOKEN_INFO is returned by C_GetTokenInfo which should return up-to-date token info. We currently supply dummy values in most cases which can be cached. But if there is a clock on a HSM, its time should prabably be displayed in CK_TOKEN_INFO's utcTime such that it can be checked against the system time for example for tampering detection.

A token which doesnt have its own clock should prabably not set CKF_CLOCK_ON_TOKEN in CK_TOKEN_INFO's flags. Only for soft token it probably makes sense to supply system time in utcTime (and maybe ica, since its some kind of soft-token).

In the best case (we guessed the right token flags), the native ep11 flags dont add anything useful. But if we get the flag information directly from the token (as intendet) why not use it?

In case of the serial number im not sure how we model the token(=one ep11lib) - adapter(possible multiple) relation.

Reply:
Regarding utcTime: OK, that would mean that we need another call to get_xcp_info to obtain the time value during C_GetTokenInfo. Here again the problem is that we can have multiple adapters. Theoretically every adapter could have a different time.... Again, I could just use the first one for simplicity.

Regarding flags: I checked and it only reports CKF_RNG, CKF_CLOCK_ON_TOKEN and CKF_TOKEN_INITIALIZED. We set those already in the code. So OR'ing the flags from the get_xcp_info call won't add anything useful. The token flags contain additional flags that the EP11 lib can not provide, such as PIN states, etc.